crates/type-system/src/ontology/shared/one_of/mod.rs

```diff
@@ -0,0 +1,66 @@
+pub mod error;
+pub(in crate::ontology) mod repr;
+
+use crate::{ParseOneOfError, ValidationError};
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct OneOf<T> {
+    possibilities: Vec<T>,
+}
+
+impl<T> OneOf<T> {
+    /// Creates a new `OneOf` without validating.
+    pub fn new_unchecked<U: Into<Vec<T>>>(possibilities: U) -> Self {
+        Self {
+            possibilities: possibilities.into(),
+        }
+    }
+
+    /// Creates a new `OneOf` from the given vector.
+    ///
+    /// # Errors
+    ///
+    /// - [`ValidationError`] if the object is not in a valid state.
+    pub fn new<U: Into<Vec<T>>>(one_of: U) -> Result<Self, ParseOneOfError> {
+        let one_of = Self::new_unchecked(one_of);
+        one_of
+            .validate()
+            .map_err(ParseOneOfError::ValidationError)?;
+        Ok(one_of)
+    }
```

Comment:
This can only return `ValidationError`, not `ParseOneOfError`

Reply:
I don't understand your comment. It returns `ValidationError` which is wrapped inside `ParseOneOfError`, as it is a variant of a problem with the `OneOf` (this is the `OneOf::new` function after all).

Reply:
OneOf::new will only construct a single variant of `ParseOneOfError`, so the code could be changed to
```rust
    pub fn new<U: Into<Vec<T>>>(one_of: U) -> Result<Self, ValidationError> {
        let one_of = Self::new_unchecked(one_of);
        one_of.validate()?;
        Ok(one_of)
    }
```